0001-job-worker.md

```diff
@@ -0,0 +1,257 @@
+
+Logs have been chosen to store on disk over memory for several reasons:
+ - Simplicity: It's easier to simply specify a file descriptor to exec.Cmd.StdOut and StdErr to concurrently write the output to a file. It's also easier to concurrently read from a file as this is natively supported in linux. If the contents were in a buffer in memory a concurrent reading would require more code and potentially require a copy of the buffer for every reader
+ - Fault tolerance: If the job worker process dies, the in memory logs would be lost. By using a file it is possible that the job worker could resume streaming a job's output
+
+### Process execution life cycle
+
+Together with this design doc there are two UML sequence diagrams showing a happy and un happy scenario.
+
+![unhappy.plantuml](unhappy.jpg "Unhappy Scenario")
+
+In the unhappy scenario a gRPC client sends a start request to the server. The server then uses the library to first create a cgroup to execute the job, adds the resource control parameters in the appropriate interface files for the cgroup controllers, then starts the job's process. This process is not the client's command, but a simple process that adds it's own PID to the cgroup. This prevents a race condition where the client's command is executed then the PID is found and added, since the command may already abuse resources before the cgroup has a chance to control them. Once the PID is added the command is run in the process. After this the server responds to the client. In this un happy scenario, we assume the command fails quickly and returns an error code. The client then uses the ID in the response to query the status of the job. Since the server has access to the underlying exec.Cmd and therefore os.Process we can return the status along with the return code to the user with the Running bool as false.
+
+![happy.plantuml](happy.jpg "Happy Scenario")
+
+The happy scenario is the same for starting a job. Here we assume the command has not yet returned and is executing, so the status response returns Running true. Therefore the client can now use the ID again to stream the logs of the job. After streaming for a period of time the client can cancel the stream. If the client then decides to stop the job it can send the ID to the server which sends a SIGTERM signal to the process, i.e. `cmd.Process.Signal(syscall.SIGTERM)`. Once stopped the cgroup is deleted and a response sent to the client.
```

Comment:
Possibly naïve question: Will the `cgroup` (or any other) cleanup have any impact on clients asking for the command output long after the job has finished?

Reply:
It depends, is there a requirement for consumers to query the command's output after the job has stopped? I was thinking when a consumer calls Close() this would clean up the cgroup and the log file. But we could instead only delete the cgroup and maintain the log file for inspecting after. We would then need another function to clean up this file, something like a Close()